Adding blocklist to move randomization for TM's and learnsets

danpatoneil opened the pull request:
Thanks for contributing to the Pokémon Crystal Archipelago project!

## What does this add?
A block list for moves to options. Filters those moves specified in the yaml out of the pool for randomized learnsets and randomized TM's.

## Why do you want it to be added?
It's a requested feature and allows for some amount of control over randomized movesets.

## Was this tested yet?
This was tested manually a few different times and ways, I tried to look at a unit test for something like this but couldn't find it in either of the previous pokemon games so I didn't write one

palex00 committed:
Update pull_request_template.md

palex00 committed:
Update bug_report.yaml

palex00 committed:
Update feature_request.yaml

palex00 committed:
Delete .github/ISSUE_TEMPLATE/task.yaml

danpatoneil committed:
added move blacklist to the options, currently does nothing

danpatoneil committed:
Merge branch 'pokecrystal' of https://github.com/danpatoneil/Archipelago-Crystal into pokecrystal

gerbiljames committed:
Merge pull request #67 from palex00/patch-12

Update PR + Issues / Feature Request Templates

danpatoneil committed:
Added move blacklist into get_random_move method

still needs to be added to tm's

danpatoneil committed:
tweaked wording in options

danpatoneil committed:
Added move blacklist functionality to TM generation

danpatoneil committed:
Merge branch 'gerbiljames:pokecrystal' into pokecrystal

danpatoneil committed:
cleaned up code slightly, consolidated

danpatoneil committed:
Merge branch 'pokecrystal' into move-block-list

danpatoneil committed:
adding spacing after merging around

danpatoneil committed:
Merge branch 'pokecrystal' into move-block-list

gerbiljames commented on worlds/pokemon_crystal/moves.py:
```diff
@@ -69,6 +73,11 @@ def get_tmhm_compatibility(world: "PokemonCrystalWorld", pkmn_name):
+    
+    #remove every move from move_pool that is in the blacklist
+    if world.options.move_blacklist:
+        move_pool = [move for move in move_pool if move.name not in [m.replace('_', ' ') for m in world.options.move_blacklist]]
```
It would be good if the comparison here was case insensitive so the YAML could contain "Dragon Rage" or "DRAGON RAGE"

gerbiljames commented on worlds/pokemon_crystal/moves.py:
```diff
@@ -40,10 +40,14 @@ def get_random_move(random, move_type=None, attacking=None):
+    
+    #remove every move from move_pool that is in the blacklist
+    if world.options.move_blacklist:
+        move_pool = [move_name for move_name in move_pool if move_name not in world.options.move_blacklist]
```
I don't think we should be expecting users to submit moves in the form "DRAGON_RAGE", they would expect to be able to use forms such as "Dragon Rage" or "DRAGON RAGE" too.

gerbiljames commented on worlds/pokemon_crystal/options.py:
```diff
@@ -463,6 +464,13 @@ class EnableMischief(Toggle):
     """
     display_name = "Enable Mischief"
 
+class MoveBlacklist(OptionSet):
```
Let's change this to `MoveBlocklist` to be as inclusive as possible with our language

gerbiljames reviewed (changes requested):
Looks good, apart from a few minor issues which I've added as comments.

Would you also be able to retarget this PR at the `future/3.1.0` branch?

danpatoneil committed:
added lowercase and spaced inputs in blocklist

changed options.py to add capability to accept inputs that were lower case or used a space instead of an underscore

danpatoneil committed:
added case and underscore fixing

added logic into move randomization to normalize inputs to the correct format, tm list has spaces and move pool has underscores

danpatoneil commented:
I pushed some updates to this, although I'm really new to using github on other people's projects so let me know if you need me to do anything else to move this process along.

My solution to the input problems in options.py isn't exactly elegant. There's probably some better way to approach that someone more familiar with the software knows, somewhere we can format the inputs before they're checked as valid keys or not?

danpatoneil committed:
changing blacklist to blocklist

gerbiljames commented:
Just need to to change the target branch to `future/3.1.0` and I'll get this merged :clap:

danpatoneil commented:
Whoops! Sorry I’m a little new to GitHub contributing. Should be correct now
